application/src/main/java/run/halo/app/theme/dialect/InjectionExcluderProcessor.java

```diff
@@ -0,0 +1,92 @@
+import org.springframework.util.Assert;
+import org.thymeleaf.context.ITemplateContext;
+import org.thymeleaf.model.ITemplateEnd;
+import org.thymeleaf.model.ITemplateStart;
+import org.thymeleaf.processor.templateboundaries.AbstractTemplateBoundariesProcessor;
+import org.thymeleaf.processor.templateboundaries.ITemplateBoundariesProcessor;
+import org.thymeleaf.processor.templateboundaries.ITemplateBoundariesStructureHandler;
+import org.thymeleaf.standard.StandardDialect;
+import org.thymeleaf.templatemode.TemplateMode;
+
+/**
+ * <p>Determine whether the current template being rendered needs to exclude the processor of
+ * code injection. If it needs to be excluded, set a local variable.</p>
+ * <p>Why do you need to set a local variable here instead of directly judging in the processor?</p>
+ * <p>Because the processor will process the fragment, and if you need to exclude the <code>login
+ * .html</code> template and the login.html is only a fragment, then the exclusion logic will
+ * fail, so here use {@link ITemplateBoundariesProcessor} events are only fired for the
+ * first-level template to solve this problem.</p>
+ *
+ * @author guqing
+ * @since 2.20.0
+ */
+public class InjectionExcluderProcessor extends AbstractTemplateBoundariesProcessor {
+
+    public static final String EXCLUDE_INJECTION_VARIABLE =
+        InjectionExcluderProcessor.class.getName() + ".EXCLUDE_INJECTION";
+
+    private final PageInjectionExcluder injectionExcluder = new PageInjectionExcluder();
+
+    public InjectionExcluderProcessor() {
+        super(TemplateMode.HTML, StandardDialect.PROCESSOR_PRECEDENCE);
+    }
+
+    @Override
+    public void doProcessTemplateStart(ITemplateContext context, ITemplateStart templateStart,
+        ITemplateBoundariesStructureHandler structureHandler) {
+        if (isExcluded(context)) {
+            structureHandler.setLocalVariable(EXCLUDE_INJECTION_VARIABLE, true);
+        }
+    }
+
+    @Override
+    public void doProcessTemplateEnd(ITemplateContext context, ITemplateEnd templateEnd,
+        ITemplateBoundariesStructureHandler structureHandler) {
+        structureHandler.removeLocalVariable(EXCLUDE_INJECTION_VARIABLE);
+    }
+
+    /**
+     * Check if the template will be rendered is excluded injection.
+     *
+     * @param context template context
+     * @return true if the template is excluded, otherwise false
+     */
+    boolean isExcluded(ITemplateContext context) {
+        return injectionExcluder.isExcluded(context.getTemplateData().getTemplate());
+    }
+
+    static class PageInjectionExcluder {
+
+        private final Set<String> exactMatches = Set.of(
+            "login",
+            "signup",
+            "logout"
+        );
+
+        private final Set<Pattern> regexPatterns = Set.of(
+            Pattern.compile("error/.*"),
+            Pattern.compile("challenges/.*"),
+            Pattern.compile("password-reset/.*"),
+            Pattern.compile("login_.*")
```

Comment:
这个规则似乎可以删除掉。目前不会**直接**渲染以 `login_` 开头的模板。一般都是通过 fragment 的方式引用，因为我们是在 TemplateBoundariesProcessor 中进行判断的，所以模板名不会包含这些 fragment template。

```suggestion
            Pattern.compile("password-reset/.*")
```

Reply:
Ok